### DISCUSSION TITLE
Issue in running the script "twice"

### DISCUSSION THREAD

**Discussion by diogo-rossi** [haiiliin/abqpy#1362] [Ideas]
Some times, there will be found another issue regarding the place where the `run()` function is called.

Imagine the following situation:

```python
# Contents of `script.py`
from abaqus import *
from abaqusConstants import *
from driverUtils import *

# Do code stuff:
# * Read and generate data and make calculations 
# * Open, create, delete, move files and folders 
# * Change the CWD to launch Abaqus in separate folder
# * etc...

mdb.saveAs('model.cae')
```

In the above code, the `run()` function is called in the `mdb.saveAs()` method, which launches Abaqus and exit.

Because of that, if `mdb.saveAs()` is called at the end, the "code stuff" will run twice: by the `python3` interpreter (with the command line `python3 script.py`) and by the `abaqus python2` interpreter. And that could cause some issues in operations like writing files and moving folders.

### One possible solution is:

Move the call of `run()` function to the modules `abaqus` and `odbAcess`.

When we `import abaqus` or `import odbAcess`, the function will launch Abaqus and exit, not running the "code stuff" in `python3`interpreter

Is that feasible?

**Comment by haiiliin** [comment]
I think when building the model, we would rarely write something, only when we are getting the output data. 

If we do this, much of the code will not be executed in python 3 interpreter, what if we want to do some debugging in python 3 to see what is in the model, though it is different in the abaqus python 2 interpreter, we can still get some information about the model.

**Reply by diogo-rossi** [reply]
Oh, I got your point, that would really be useful.

But still: what if I want to do several launches and buildings with one single script? (I mean, create folders, moving to folders, etc)

Another thing is: the script will behave differently in `python3` environment and in `abaqus python2` interpreter, which may be inconsistent, as the point is to write the same script in an different environment, I think.

> what if we want to do some debugging in python 3 to see what is in the model

That would be really useful! And I think [this is a solution for that](https://stackoverflow.com/questions/38634988/check-if-program-runs-in-debug-mode).

**Reply by haiiliin** [reply]
Oh, this should be useful.

**Reply by diogo-rossi** [reply]
With that, maybe we don't need to use `exit_after` argument in `run()` function. It could decide whether or not to exit (if in debuging or in normal run)

**Reply by haiiliin** [reply]
Right, excellent idea.

**Comment by diogo-rossi** [comment]
I was think about to move all the calls of the `abqpy.abaqus.run()` to only 2 places:
___
* End of file `src/abaqus/___init__.py`:
```python
...
gettrace = getattr(sys, 'gettrace', None)
exit_after = False
if gettrace is None or not gettrace():
    exit_after = True

abqpy.abaqus.run(cae=True, exit_after=exit_after)

```
* End of file `src/odbAccess.py`:
```python
...
gettrace = getattr(sys, 'gettrace', None)
exit_after = False
if gettrace is None or not gettrace():
    exit_after = True

abqpy.abaqus.run(cae=False, exit_after=exit_after)
```
___
In this way, both `import abaqus` and `import odbAccess` will run the `abaqus` command and exit in Python 3 interpreter
So, it will not run the rest of the script twice (because it will already run in Python 2).
Except for when in debug mode. In this case, Python 3 will run the file anyway.

What do you think?

**Reply by haiiliin** [reply]
I totally agree with you

**Reply by diogo-rossi** [reply]
I've made the changes in [this branch](https://github.com/haiiliin/abqpy/tree/way-to-not-run-the-script-twice). Do you think it is ok?

**Reply by haiiliin** [reply]
Have you tried to run it in debug mode? is it working?

**Reply by haiiliin** [reply]
I think the `OdbCommands` also need to be updated.

**Reply by diogo-rossi** [reply]
Yes, I'll check that

**Reply by haiiliin** [reply]
https://github.com/haiiliin/abqpy/blob/e74cbe434c7ff17a3d1cebef94e468ce1eb3d14f/src/odbAccess.py#L7-L13

How about moving all this to `abqpy.abaqus.run`?

**Reply by diogo-rossi** [reply]
Makes sense.

**Reply by diogo-rossi** [reply]
I was having an `AssertionError` here, and hence made this modification (on line `26`)

I did not understand very well this error.

https://github.com/haiiliin/abqpy/blob/a5125065f0c6b25a826e703a61c33d06c7ce6d6d/src/abqpy/__init__.py#L19-L27

**Reply by haiiliin** [reply]
How about adding an `Exception` to catch all kinds of errors?

**Reply by diogo-rossi** [reply]
Ok. Do you understand why this error now? I wasn't having it before.

**Reply by haiiliin** [reply]
I actually do not know this happens why, either.

**Reply by haiiliin** [reply]
I am using `pip install .` to install, nothing happens

**Reply by diogo-rossi** [reply]
Now the scripts are running OK, but I could not handle the `SystemExit` error that happens with `pytest`.

The only way I found in how to handle with `pytest` is the solution below. Although I think it is not a good solution. 

https://github.com/haiiliin/abqpy/blob/71d004134804915f72ce0dcd0ace4b35628718db/src/abqpy/abaqus.py#L39-L40

**Reply by haiiliin** [reply]
How about moving all the imports under `pytest.raise`

**Reply by haiiliin** [reply]
We can define a environment variable to indicate whether to exit in `abaqus` and `odbAccess` module, the variable is True by default, in pytest, we set the variable to False.

**Reply by diogo-rossi** [reply]
> How about moving all the imports under `pytest.raise`

I tried that, but the test failed.

It seems that the `exit` makes `pytest` continue the script, but cancels the imports.

So, further in the script, it raises an `NameError`

**Reply by haiiliin** [reply]
> We can define a environment variable to indicate whether to exit in `abaqus` and `odbAccess` module, the variable is True by default, in pytest, we set the variable to False.

This should be possible

**Reply by diogo-rossi** [reply]
So, how to set the variable to `False` only in `pytest`?

**Reply by haiiliin** [reply]
Possibly in the `pytest_configure` function, use `os.environ` to set environment variable

**Reply by haiiliin** [reply]
https://github.com/haiiliin/abqpy/blob/84f7eec2598386534b29720901b0e4c6450b0653/src/abqpy/abaqus.py#L33-L43
https://github.com/haiiliin/abqpy/blob/84f7eec2598386534b29720901b0e4c6450b0653/tests/conftest.py#L9-L21

This should solve the problem

**Reply by haiiliin** [reply]
https://github.com/haiiliin/abqpy/blob/e7ed5b21b0ee78664f8d6d7c54785a4902f67c2f/src/abqpy/abaqus.py#L29-L41

And I think we should not run the Abaqus command at all if it is in debug mode.

**Comment by haiiliin** [comment]
Implemented in #1501.
